mm/mempool/mempool_multiple.c

```diff
@@ -220,7 +220,7 @@ mempool_multiple_get_dict(FAR struct mempool_multiple_s *mpool,
   col = index - (row << mpool->dict_col_num_log2);
   if (mpool->dict[row] == NULL ||
       mpool->dict[row][col].addr != addr ||
-      blk - addr >= mpool->dict[row][col].size)
+      (FAR char *)blk - (FAR char *)addr >= mpool->dict[row][col].size)
```

Comment:
Just an alternative is `uintptr_t`.

Reply:
yes, but then you'd need to do something like (ssize_t)((uintptr_t)blk - (uintptr_t)addr) . otherwise the sign of the calculation may be wrong (never negative but overflown to large positive) as it is compared against unsigned. and this format just looks even uglier (and relies on the fact(?) that sizeof(ssize_t) == sizeof(uintptr_t). which is maybe so, but not by standard (ssize_t just needs to be large enough to cover the whole address area, there is no requirement for it to be exactly the same size... *if* it would be larger this would fail still....  Just keeping pointers as pointers when doing pointer arithmetic is IMHO  cleanest, and then you can also trust that the result is proper signed integer :)

Reply:
@jlaitine which compiler are you using? Look like your compiler generate more warning than the official CI.

Reply:
I am only using gcc, version depending on the target (different versions for x86, arm linux, arm embedded, SiFive's risc compilers). But the issue here is just that the compiler flags in NuttX CI doesn't enable warnings for non-standard gcc extensions. It might be a good idea to go through the default compiler settings and set warnings on for non-standard extensions (+ maybe use -Werror in addition).

A strictly standard following flags would be "-ansi -pedantic -Wall". Just to enable warning on void * arithmetic (which is btw forbidden in standards, but still has the gcc extension which treats it the same as arithmetic on char *) can be enabled individually in gcc with -Wpointer-arith

Since I am working on PX4, and they use by default also -Werror (which I think is a good idea, warnings might go unnoticed otherwise), fixing these is important...

Reply:
Will `intptr_t` solve sign issue?

Reply:
sure it would! Cleanest solution is always first cast the void * to the real target type ptr and then do arithmetic. This also gives the compiler the opportunity to know the maximum result size. intptr_t usage is IMHO useless almost always, and just think about it's bit size (it is minimum the same size of pointer + 1 bit). It also doesn't exist on all platforms (not sure if posix enfoces the existence though, not going to check it up now...).

With pointers it is always safe and portable to do pointer arithmetic, but not on void * since by the standard you can't (obviously because of size of void). But if you don't know the target type, you always get the same result as gcc extension by casting to char *. So changing void * arithmetic by casting the pointers to char * doesn't change the existing functionality.